BF install

bpoldrack opened the pull request:
Closes #883 
Closes #944

yarikoptic commented on datalad/distribution/tests/test_install.py:
```diff
@@ -501,3 +510,47 @@ def test_reckless(path, top_path):
+    top_src.save(auto_add_changes=True, recursive=True)
+
+    # install top level:
+    top_ds = install(path=path, source=src)
```
we could also test cmdline API by simply calling `datalad.cmdline.main:main(['datalad', 'install', '--source', ...]`

bpoldrack replied:
Didn't do it that way, but good point in general!

yarikoptic commented on datalad/distribution/install.py:
```diff
@@ -309,34 +309,42 @@ def __call__(
+                            # Note: We don't exactly know what was skipped but
+                            # the `path` requested to be installed, since it will be
+                            # resolved only within the recursive call of install.
+                            lgr.info("Installation of {0} skipped.".format(p))
```
add exception itself into the message (`exc_str`)

yarikoptic replied:
but also -- make test trigger it

bpoldrack replied:
Will not include the exception, since it is logged before. This would be redundant. It's just about communicating, that the exception that just occured and was logged led to skip installation of this particular `path` argument and we proceed we the next one.

But yes - a test should trigger it.

yarikoptic replied:
again -- you might not know what is logged or not before when you catch this exception.  IMHO For now - better double logging than trying to figure out later WTF.
Thought about this skipping business anyways -- and thought that by default we shouldn't skip anything and only if explicit "--skip-failing" (or --failing=skip), then proceed as if it is ok

yarikoptic commented on datalad/distribution/install.py:
```diff
@@ -725,6 +731,10 @@ def __call__(
 
+                except Exception:
+                    # Error itself should already be logged.
+                    lgr.info("{0} skipped.".format(subds))
```
same here  -- include exception

bpoldrack replied:
Nope. For the same reason as before. This is dealt with in the recursive instance. It's just about skipping.

yarikoptic replied:
I got that -- but you cannot guarantee that you know from this msg what happened to make it skip.  That is why it never hurts to add at least some detail about the exception.  I already had to troubleshoot many of such points where no clear reason was logged and then things were just skipped.

yarikoptic commented:
already in conflict with #959 I guess

bpoldrack commented:
Yes, it conflicts. Added a test (and changed another one) to better cover the case of lists and trigger the skipping of failed installations, but will stop here without a decision on #959.
Nevertheless, could be merged from my point of view.

mih commented:
This is no need to wait for #959 

I will keep sucking in any tests that appear in this PR -- those will not conflict. The tested code might, but it will be easy to resolve.

yarikoptic commented:
I don't think that this one should wait for #959 (with yet unclear purpose, no description on purpose).  This one is fixing a bug in a release so destined for a bugfix release,  and #959 is breaking API so destined for new minor release.

bpoldrack commented:
Agree. The fixing should not wait, so merge. There should be no uncovered line by now.
What has to wait, I guess, is working on things like exit code and the like we talked about. Because these are general behaviour issues, that involve several commands and most likely will conflict.

yarikoptic commented on datalad/distribution/tests/test_install.py:
```diff
@@ -570,6 +570,33 @@ def test_install_list_arguments(src, path, path_outside):
+    sub2 = Dataset(opj(path, 'subm 2'))
+    # sabotage recursive installation of 'subm 1' by polluting the target:
+    with open(opj(path, 'subm 1', 'blocking.txt'), "w")  as f:
+        f.write("sdfdsf")
```
Just FYI I usually use `create_tree` (parent of `@with_tree`) when I need to create some files/dirs for testing
